src/dynamics/implicit.jl

```diff
@@ -225,65 +292,359 @@ function initialize!(
+    implicit.U .= on_architecture(arch, U_cpu)
+    implicit.W .= on_architecture(arch, W_cpu)
+    implicit.L0 .= on_architecture(arch, L0_cpu)
+    implicit.L1 .= on_architecture(arch, L1_cpu)
+    implicit.L2 .= on_architecture(arch, L2_cpu)
+    implicit.L3 .= on_architecture(arch, L3_cpu)
+    implicit.L4 .= on_architecture(arch, L4_cpu)
 end
 
 set_initialized!(implicit::ImplicitPrimitiveEquation) = (implicit.initialized = true)
 
 """$(TYPEDSIGNATURES)
 Apply the implicit corrections to dampen gravity waves in the primitive equation models."""
-function implicit_correction!(
+function implicit_correction_kernels!(
+    diagn::DiagnosticVariables,
+    progn::PrognosticVariables,
+    implicit::ImplicitPrimitiveEquation,
+    model::PrimitiveEquation,
+)
+    # escape immediately if explicit
+    implicit.α == 0 && return nothing
+
+    (; nlayers, trunc) = implicit
+    (; S⁻¹, R, U, L, W) = implicit
+    ξ = implicit.ξ[]
+
+    (; temp_tend, pres_tend, div_tend) = diagn.tendencies
+    div_old, div_new = get_steps(progn.div)
+    G = diagn.dynamics.a              # reuse work arrays, used for combined tendency G
+    geopotential = diagn.dynamics.b   # used for geopotential
+    geopotential .= 0
+
+    # Get precomputed l_indices from the spectrum
+    l_indices = temp_tend.spectrum.l_indices
+
+    arch = architecture(temp_tend)
+    
+    # MOVE THE IMPLICIT TERMS OF THE TEMPERATURE EQUATION FROM TIME STEP i TO i-1
+    # geopotential and linear pressure gradient (divergence equation) are already evaluated at i-1
+    # so is the -D̄ term for surface pressure in tendencies!
+
+    # SEMI IMPLICIT CORRECTIONS FOR DIVERGENCE
+    # calculate the combined tendency G = G_D + ξRG_T + ξUG_lnps to solve for divergence δD
+
+    # Kernel 1: Temperature update and geopotential calculation
+    launch!(arch, SpectralWorkOrder, size(temp_tend), 
+            implicit_temp_geopot_kernel!, 
+            temp_tend, geopotential, div_old, div_new, R, L, nlayers)
+    
+    # Kernel 2: Calculate G = G_D + ξRG_T + ξUG_lnps
+    launch!(arch, SpectralWorkOrder, size(div_tend),
+            implicit_calculate_G_kernel!,
+            G, div_tend, geopotential, pres_tend, U, l_indices, ξ)
+    
+    # Kernel 3: Solve for divergence δD = S⁻¹G
+    launch!(arch, SpectralWorkOrder, size(div_tend),
+            implicit_solve_divergence_kernel!,
+            div_tend, G, S⁻¹, l_indices, nlayers)
+    
+    # Kernel 4: Correct temperature
+    launch!(arch, SpectralWorkOrder, size(temp_tend),
+            implicit_correct_temp_kernel!,
+            temp_tend, div_tend, L, ξ, nlayers)
+    
+    # Kernel 5: Correct pressure (parallelize only over lm, not k)
+    lm_size = size(pres_tend, 1)
+    launch!(arch, SpectralWorkOrder, (lm_size,),
+            implicit_correct_pres_kernel!,
+            pres_tend, div_tend, W, ξ, nlayers)
+
+    zero_last_degree!(div_tend)
+    zero_last_degree!(pres_tend)
+    zero_last_degree!(temp_tend)
```

Comment:
Also `zero_last_degree!(vor_tend)` while we're at it? Then the `spectral_truncation!` in `leapfrog!` can go too!